sui_types/src/waypoint.rs

```diff
@@ -0,0 +1,445 @@
+pub struct WaypointWithItems<K, I>
+where
+    K: Clone,
+    I: Ord,
+{
+    pub key: K,
+    pub waypoint: Waypoint,
+    pub items: BTreeSet<I>,
+}
+
+impl<K, I> WaypointWithItems<K, I>
+where
+    K: Clone,
+    I: Borrow<[u8]> + Ord,
+{
+    pub fn new(key: K, sequence_number: u64) -> WaypointWithItems<K, I> {
+        WaypointWithItems {
+            key,
+            waypoint: Waypoint::new(sequence_number),
+            items: BTreeSet::new(),
+        }
+    }
+
+    /// Insert an element in the accumulator and list of items
+    pub fn insert_full(&mut self, item: I) {
+        self.waypoint.accumulator.insert(&item);
+        self.items.insert(item);
+    }
+
+    /// Insert an element in the accumulator only
+    pub fn insert_accumulator(&mut self, item: &I) {
+        self.waypoint.accumulator.insert(item);
+    }
+
+    /// Insert an element in the items only
+    pub fn insert_item(&mut self, item: I) {
+        self.items.insert(item);
+    }
+}
+
+/*
+    Represents the difference between two waypoints
+    and elements that make up this difference.
+*/
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct WaypointDiff<K, I>
+where
+    K: Clone,
+    I: Borrow<[u8]> + Ord,
+{
+    pub first: WaypointWithItems<K, I>,
+    pub second: WaypointWithItems<K, I>,
+}
+
+impl<K, I> WaypointDiff<K, I>
+where
+    K: Clone,
+    I: Borrow<[u8]> + Ord,
+{
+    pub fn new<V1, V2>(
+        first_key: K,
+        first: Waypoint,
+        missing_from_first: V1,
+        second_key: K,
+        second: Waypoint,
+        missing_from_second: V2,
+    ) -> WaypointDiff<K, I>
+    where
+        V1: Iterator<Item = I>,
+        V2: Iterator<Item = I>,
+    {
+        let w1 = WaypointWithItems {
+            key: first_key,
+            waypoint: first,
+            items: missing_from_first.collect(),
+        };
+        let w2 = WaypointWithItems {
+            key: second_key,
+            waypoint: second,
+            items: missing_from_second.collect(),
+        };
+
+        WaypointDiff {
+            first: w1,
+            second: w2,
+        }
+    }
+
+    /// Swap the two waypoints.
+    pub fn swap(self) -> WaypointDiff<K, I> {
+        WaypointDiff {
+            first: self.second,
+            second: self.first,
+        }
+    }
+
+    /// Check the internal invarients: ie that adding to both
+    /// waypoints the missing elements makes them point to the
+    /// accumulated same set.
+    pub fn check(&self) -> bool {
+        let mut first_plus = self.first.waypoint.accumulator.clone();
+        first_plus.insert_all(self.first.items.iter());
+
+        let mut second_plus = self.second.waypoint.accumulator.clone();
+        second_plus.insert_all(self.second.items.iter());
+
+        first_plus == second_plus
+    }
+}
+
+/*
+    A global checkpoint is the collection of differences
+    fully connecting 2f+1 by stake authorities, with diffs
+    that can be derived between all of them.
+*/
+#[derive(Clone)]
```

Comment:
Maybe separate the files for the accumulator and for using the accumulator as a commitment/succinct representation to checkpoints?

Reply:
I significantly reduced the scope of the accumulator now that it does not handle seq numbers